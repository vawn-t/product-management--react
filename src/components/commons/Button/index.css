.btn {
  border-radius: 3px;
  cursor: pointer;
  font-size: 16px;
  letter-spacing: 0.03rem;
  padding: 0.5rem 1.5rem;
}

.btn:hover,
.btn-disabled {
  opacity: 0.5;
}

/* - Default Button */
/* - ---------------------------------------------- */
.btn-default {
  background-color: var(--white);
  border: 1px solid var(--primary);
  color: var(--dark);
}

/* - Primary Button */
/* - ---------------------------------------------- */
.btn-primary {
  background-color: var(--primary);
  border: 1px solid var(--primary);
  color: var(--white);
}

/* - Secondary Button */
/* - ---------------------------------------------- */
.btn-secondary {
  background-color: var(--danger);
  border: 1px solid var(--danger);
  color: var(--white);
}

/* - Disable Button */
/* - ---------------------------------------------- */
.btn-disabled {
  cursor: auto;
}
