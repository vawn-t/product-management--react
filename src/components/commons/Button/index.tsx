// Libraries
import React, { memo, ReactNode } from 'react';

// Types
import { ButtonVariants } from '@common-types/index';

// Styles
import './index.css';

export interface ButtonProps {
  children: ReactNode;
  variant?: ButtonVariants;
  isDisabled?: boolean;
  onClick: () => void;
}

const Button: React.FC<ButtonProps> = ({
  children,
  variant = ButtonVariants.Default,
  isDisabled = false,
  onClick
}) => (
  <button
    className={`btn btn-${variant}${isDisabled ? ' btn-disabled' : ''}`}
    disabled={isDisabled}
    onClick={onClick}
  >
    {children}
  </button>
);

export default memo(Button);
