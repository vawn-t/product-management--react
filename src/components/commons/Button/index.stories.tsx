// Libraries
import React, { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import Button from './index';

// Styles
import '@assets/styles/App.css';
import '@assets/styles/variables.css';

// Types
import { ButtonVariants } from '@common-types/index';

export default {
  title: 'Components/Button',
  component: Button
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: 'Button'
};

export const Primary = Template.bind({});
Primary.args = {
  children: 'Button',
  variant: ButtonVariants.Primary
};

export const Secondary = Template.bind({});
Secondary.args = {
  children: 'Button',
  variant: ButtonVariants.Secondary
};

export const Disabled = Template.bind({});
Disabled.args = {
  children: 'Button',
  isDisabled: true
};
