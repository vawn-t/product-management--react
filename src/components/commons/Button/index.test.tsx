// Libraries
import React from 'react';
import { render, act, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';

// Components
import Button, { ButtonProps } from './index';

describe('Button component', () => {
  const defaultProps: ButtonProps = {
    children: 'Button',
    onClick: jest.fn()
  };
  let container: HTMLElement;
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultTree = create(<Button {...defaultProps} />).toJSON();
    expect(defaultTree).toMatchSnapshot();

    const disableTree = create(<Button {...defaultProps} isDisabled />);
    expect(disableTree).toMatchSnapshot();
  });

  test('should render button with content is Button', () => {
    act(() => {
      render(<Button {...defaultProps} />, { container });
    });

    expect(container.textContent).toMatch('Button');
  });

  test('should be click 2 times', () => {
    const myMock = jest.fn();
    render(<Button {...defaultProps} onClick={myMock} />, { container });

    const button: HTMLElement = screen.getByRole('button');

    fireEvent.click(button);

    expect(myMock).toHaveBeenCalled();
  });
});
