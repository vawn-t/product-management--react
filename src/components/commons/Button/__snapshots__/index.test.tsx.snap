// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render correctly 1`] = `
<button
  className="btn btn-default"
  disabled={false}
  onClick={[MockFunction]}
>
  Button
</button>
`;

exports[`Button component should render correctly 2`] = `
<button
  className="btn btn-default btn-disabled"
  disabled={true}
  onClick={[MockFunction]}
>
  Button
</button>
`;
